walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -56,4 +54,85 @@
 
     return validTimes;
   }
+
+  /** 
+    * Returns a collection of times that work for an optimal number of optional 
+    * people, as well as all the mandatory people.  
+    */
+  private Collection<TimeRange> getOptimalSubsetTimes(
+    Set<String> mandatoryAttendees, Set<String> optionalAttendees, 
+    Collection<Event> events, MeetingRequest request) {
+    
+    Set<Set<String>> subsets = Sets.powerSet(optionalAttendees);
+    Set<Set<String>> infeasibleGroups = new HashSet<Set<String>>();
+    Collection<TimeRange> optimalListOfTimes = new ArrayList<TimeRange>();
+
+    for (int i = 1; i < optionalAttendees.size(); i++) {
+      int currentSizeOfSubsetsToCheck = i;
+      Set<Set<String>> subsetsToCheck = Sets.filter(subsets, subset -> 
+        subset.size() == currentSizeOfSubsetsToCheck && 
+          !containsInfeasibleGroup(subset, infeasibleGroups));
+      if (subsetsToCheck.isEmpty()) {
+        // all subsets larger than this will be infeasible 
+        break;
+      }
+
+      for (Set<String> subset : subsetsToCheck) {
+        Collection<TimeRange> possibleTimes = 
+          getPossibleTimes(events, Sets.union(mandatoryAttendees, subset), request);
+        if (possibleTimes.isEmpty()) {
+          // for any group of optional people containing this group, no meeting 
+          // can be scheduled.
+          infeasibleGroups.add(subset);
+        } else {
+          optimalListOfTimes = possibleTimes;
+        }
+      }
+    }
+
+    if (optimalListOfTimes.isEmpty() && !mandatoryAttendees.isEmpty()) {
+      return getPossibleTimes(events, mandatoryAttendees, request);
+    } else {
+      return optimalListOfTimes;
+    }
+
+  }
+
+  /** 
+    * Returns true if `set` contains a group of people that has already been 
+    * determined to have no mutually open timeslots between the group members 
+    * and the mandatory meeting participants. 
+    */
+  private boolean containsInfeasibleGroup(Set<String> set, 
+    Set<Set<String>> infeasibleGroups) {
+    return infeasibleGroups.stream().reduce(false, (accumulator, infeasibleGroup) -> {
+      return accumulator || set.containsAll(infeasibleGroup);
+    }, Boolean::logicalOr);
+  }
+
+  /** 
+    * If possible, returns all possible time ranges in which the mandatory and 
+    * optional attendees of the requested meeting could meet for the requested 
+    * duration, without conflicting with any `events` that are already scheduled 
+    * for these attendees. If no times are possible that work for all the attendees,
+    * returns possible times such that all mandatory attendees and as many 
+    * optional antendees as possible can attend the meeting at these times. 
+    */
+  public Collection<TimeRange> query(Collection<Event> events, 
```

Comment:
It's more common to put public methods at the top of all methods and let the readers decide whether to jump into private methods.

Reply:
Noted, will fix in my next commit!